mock-array-big: README.md on Element/constraint.sdc (#1151)
The-OpenROAD-Project/OpenROAD-flow: flow/designs/asap7/mock-array-big/README.md

```
make
```

Element/constraint.sdc
----------------------

A single macro is made for all the elements in the array, though strictly
speaking each macro is unique in that the timing constraints of each element
in the array is different.

To illustrate, consider the `assign io_lsbOuts_0 = io_lsbIns_1;` statement in Element.v.
Here a signal is routed from left to right, through the Element without the signal
being registered.

Clearly, the maximum input delay for `io_lsbIns_1` is
smaller for the leftmost than for the rightmost Element.

Taking a step back, it is also worth considering what level of detail is
appropriate for the Element's constraint.sdc file.

In the beginning of a project, during exploration, the .sdc file does not

> need to be particularly detailed. In fact, details can be counterproductive.

Reviewer: Missing a "why" here so I threw this in:

```suggestion
need to be particularly detailed. In fact, details can be counterproductive, as they tend to be inaccurate and cause the tools to spend time solving potential non-issues.
```

Reviewer 2: Each step within the implementation flow will be related to different timing "magnitude/concerns", I think we need to segregate (separate by concerns) between different timing concerns within each step of the flow and/vs the tools accuracy.

Consider minimum input delay, which relates to hold times. If the main concern
of architectural exploration is to ensure that the design can operate on a
sufficiently high frequency, then specifying minimum input delays during
exploration is premature. Minimum input delay is a constraint that belongs at
the end the development cycle when the design is well established and won't change

> and the concern is to lock down the macro for it's specific operating frequency.

Reviewer: ```suggestion
and the concern is to lock down the macro for its specific operating frequency.
```

> and the concern is to lock down the macro for it's specific operating frequency.

Reviewer 2: If we are going to postpone specific timing details per each CE (with regards to its logical structure/content, position within the array and its runtime functionality) and using a generic/symmetric macro per CE, then i suggest to give the tools, at the moment, the freedom for analyzing the comb. path (within the CE) and give the its best results per CE.

> and the concern is to lock down the macro for its specific operating frequency.

Reviewer 2: If we are going to postpone specific timing details per each CE (with regards to its logical structure/content, position within the array and its runtime functionality) and using a generic/symmetric macro per CE, then i suggest to give the tools, at the moment, the freedom for analyzing the comb. path (within the CE) and give the its best results per CE.

Choosing a strategy for Element/constraint.sdc combinational paths
------------------------------------------------------------------

It isn't possible to articulate a single .sdc file that exactly captures the
constraints of an Element as each element is used in unique circumstances.

> Instead, a choice has to made and each choice has it's pros and cons.

Reviewer: ```suggestion
Instead, a choice has to made and each choice has its pros and cons.
```

Reviewer 3: what are the cons against using a unique SDC file for each element ?

It is relatively easy to set up a maximum input/output delay for paths
that start or end in a register, so this case is not discussed here.

> It is the combinational(unregistered)

Reviewer: ```suggestion
It is the combinational(unregistered) in -> out
```

paths that are tricky to create constraints for.

> - Strategy used as of writing: mark non-registered/combinational paths through

Reviewer 3: Timing will be checked but not optimized !!!
'Over- constrained'

>   the Element as false paths. This will not steer the tools to optimize the

Reviewer 2: So are we keeping the false paths?

  through element path, such as discussed above and there's no timing information
  available at the Element level for these combinational paths.
  However, at the `MockArray` level when the Elements are used, timing will
  be checked, because the .lib file generated for the Element contains timing
  information, even for false paths.

> - Use `set_max_delay` for combinational paths. This will make the tool try

Reviewer 3: it's something that can be measured by the tool
whether it helps the tool to meet better timing or not

  to optimize the combinational path and there will be timing information available.

>   However, path segmentation could occur with OpenSTA: https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation

Reviewer 4: ```suggestion
  However, path segmentation could occur with OpenSTA. Refer [here](https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation).
```

>   However, path segmentation could occur with OpenSTA. Refer [here](https://docs.xilinx.com/r/2020.2-English/ug906-vivado-design-analysis/TIMING-13-Timing-Paths-Ignored-Due-to-Path-Segmentation).

Reviewer 2: FMO we should avoid (as possible) using FPGA references with regards to ASIC issues.

- Overconstrain: set up a maximum input/output path for the element and ignore
  timing violations at the Element level for combinational paths.

>   The timing violations for the combinational paths are not real violations,

Reviewer 3: It's same incentive of defining the path as 'False path' (there is a strict definition)
Is it a not-real violation ?
what happens when the combinational paths is more complex? (more propagation delay)

Author: I don't have answers to these questions. I'm just summarizing different approaches here. If someone can research and answer these questions, then an update to this README.md is welcome.

>   The timing violations for the combinational paths are not real violations,

Reviewer 2: How does using over constraints initially can help for identify real issues with regards to FE/BE/Tools?

  because the maximum input and output paths can not occur at the same time
  in the array. The leftmost Element has the shortest maximum input path and

>   and the rightmost the longestm aximum input path for

Reviewer 4: ```suggestion
  and the rightmost the longest maximum input path for
```

  `assign io_lsbOuts_0 = io_lsbIns_1;`. The OpenROAD tool could put too much
  emphasis on optimising the combinational path in this case, which could

>   lead to less than optimal results.

Reviewer: ```suggestion
  lead to less than optimal results and/or inflated run times.
```
